pkg/objstore/oci/oci.go

```diff
@@ -0,0 +1,335 @@
+	if err != nil {
+		if b.IsObjNotFoundErr(err) {
+			return false, nil
+		}
+		return false, errors.Wrapf(err, "cannot get OCI object '%s'", name)
+	}
+	return true, nil
+}
+
+// Delete removes the object with the given name.
+// If object does not exists in the moment of deletion, Delete should throw error.
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	return deleteObject(ctx, *b, name)
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	failure, isServiceError := common.IsServiceError(err)
+	if isServiceError {
+		k := failure.GetHTTPStatusCode()
+		match := k == http.StatusNotFound
+		level.Debug(b.logger).Log("msg", match)
+		return failure.GetHTTPStatusCode() == http.StatusNotFound
+	}
+	return false
+}
+
+// ObjectSize returns the size of the specified object.
+func (b *Bucket) ObjectSize(ctx context.Context, name string) (uint64, error) {
+	response, err := getObject(ctx, *b, name, "")
+	if err != nil {
+		return 0, err
+	}
+	return uint64(*response.ContentLength), nil
+}
+
+// Close closes bucket.
+func (b *Bucket) Close() error {
+	return nil
+}
+
+// Attributes returns information about the specified object.
+func (b *Bucket) Attributes(ctx context.Context, name string) (objstore.ObjectAttributes, error) {
+	response, err := getObject(ctx, *b, name, "")
+	if err != nil {
+		return objstore.ObjectAttributes{}, err
+	}
+	return objstore.ObjectAttributes{
+		Size:         *response.ContentLength,
+		LastModified: response.LastModified.Time,
+	}, nil
+}
+
+// createBucket creates bucket.
+func (b *Bucket) createBucket(ctx context.Context, compartmentId string) (err error) {
+	return createBucket(ctx, b, compartmentId)
+}
+
+// deleteBucket deletes bucket.
+func (b *Bucket) deleteBucket(ctx context.Context) (err error) {
+	return deleteBucket(ctx, b)
+}
+
+// NewBucket returns a new Bucket using the provided oci config values.
+func NewBucket(logger log.Logger, ociConfig []byte) (*Bucket, error) {
+	level.Debug(logger).Log("msg", "creating new oci bucket connection")
+	var config = DefaultConfig
+	var configurationProvider common.ConfigurationProvider
+	var err error
+
+	if err := yaml.Unmarshal(ociConfig, &config); err != nil {
+		return nil, errors.Wrapf(err, "unable to unmarshal the given oci configurations")
+	}
+
+	provider := Provider(strings.ToLower(config.Provider))
+	level.Info(logger).Log("msg", fmt.Sprintf("Creating OCI default '%s' provider", provider))
+	switch provider {
+	case DefaultConfigProvider:
+		configurationProvider = common.DefaultConfigProvider()
+	case InstancePrincipalConfigProvider:
+		configurationProvider, err = auth.InstancePrincipalConfigurationProvider()
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to create OCI instance principal config provider")
+		}
+	case RawConfigProvider:
+		if err := config.validateConfig(); err != nil {
```

Comment:
Do we need to call this in all cases? If not, do we need extra validation for other providers?

Reply:
InstancePrincipalConfigProvider 
The OCI Instance Principe setup is when you setup the OCI objectstorage bucket. You grant the permission through OCI policies. All the checking with by covered by auth.InstancePrincipalConfigurationProvider. A nice error message will be thrown if the instance principal policy configuration has issue. There  is no need to duplicate the same logic that OCI GO SDK already covered.
See https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm.

DefaultConfigProvider
This is the most standard way to connect to OCI especially for users run stuff from their Mac (local marine). Typically, it uses the local .oci/config file to configuration the OCI provider. The common.DefaultConfigProvider() also provided a very nice error message. There  is no need to duplicate the same logic that OCI GO SDK already covered.

RawConfigProvider
Most of the OCI user don't use provider because this API requires you to pass few parameters (such as Fingerprint). This provider is for the user who want to create the OCI provider on the fly which is the vey common (and secure) in OCI world. My `validateConfig()` only covers the most basic one to remind the users what are the minimum requirements. My thinking it would be nice to spell it out because as you can see the `common.NewRawConfigurationProvider()` call requires some parameters (unlike `common.DefaultConfigProvider()` and `auth.InstancePrincipalConfigurationProvider()` which do not require any parameter). I know OCI GO SDK does provide some good error message.

If you want to make all 3 provider looks alike, I think I just remove the config.validateConfig() completely.
```
if err := config.validateConfig(); err != nil {
			return nil, errors.Wrapf(err, "invalid oci configurations")
		}
```

Reply:
Please let me know if you still have concern with this implementation.

Reply:
Thank you for this really thorough explanation, looks good!